### ISSUE TITLE
Better Annotation Handling for Return types and Type Arguments

### ISSUE THREAD

**Issue by andriy-dmytruk** [micronaut-projects/micronaut-core#7686]
### Issue description

Annotating type arguments and return types during build time (e.g. `InjectProcessor`) is currently not supported in core. The `ClassElement` adds annotations to the whole class instead of the types (usages of this class).

**Use case**: I want to annotate types whenever their type arguments have validation annotations. For example `List<@Valid B>` -> `@Valid List<@Valid B>` in order to reduce validation time during runtime. Otherwise all the parameters with type arguments would have to be checked every time, which doesn't seem desirable.

## Reproducing issue

Please check these tests: [ClassElementAnnotationValidationSpec](https://github.com/andriy-dmytruk/micronaut-core/blob/class-element-annotations/inject-java-test/src/test/groovy/io/micronaut/inject/visitor/ClassElementAnnotationValidationSpec.groovy). The first of them is run successfully while the second one fails on the last line.

## Description

Type arguments are presented as instances of `JavaClassElement`. When the [annotate()](https://github.com/andriy-dmytruk/micronaut-core/blob/db1f42eea1f305898cdfaa1244612ec9be92144d/inject-java/src/main/java/io/micronaut/annotation/processing/visitor/AbstractJavaElement.java#L87-L101) method is called for them, in turn [updateMetadataCaches()](https://github.com/andriy-dmytruk/micronaut-core/blob/db1f42eea1f305898cdfaa1244612ec9be92144d/inject-java/src/main/java/io/micronaut/annotation/processing/visitor/AbstractJavaElement.java#L167-L169) is called from `AbstractJavaElement`. The method then stores the metadata inside `AbstractAnnotationMetadataBuilder` with the `element` as key (which in this case is the java `TypeElement` not its usage and therefore annotation is added to the whole class).

Note that the first test passes since when created on ([L359](https://github.com/andriy-dmytruk/micronaut-core/blob/db1f42eea1f305898cdfaa1244612ec9be92144d/inject-java/src/main/java/io/micronaut/annotation/processing/visitor/AbstractJavaElement.java#L359)) the type arguments both receive the `EmptyAnnotationMetadata`, which is immutable. It would fail if the String class had annotations and thus they would both receive the same Object of `MutableAnnotationMetadata`. Eliminating the need to store annotations for both the type and its usage could simplify the code and behavior in general.

## Solutions

* Store the return type or type argument as `element` in `ClassElement` instead of the class. Therefore all the annotations would be applied to them, but class could not be annotated.
* Create classes such as `ReturnTypeElement` and `TypeArgumentElement` similarly to the `ParameterElement`.
* Repurpose the `ParameterElement` to accomodate to these cases  (it seems like simply renaming it more generally and using it would work).

@graemerocher @yawkat @dstepanov

**Comment by dstepanov** [comment]
I don't see any `@MyAnnotation` in the example

**Comment by graemerocher** [comment]
yeah the tests don't see to reflect the behaviour you are interested in.

Not saying this is not an issue since it is, but also if the use case is to annotate to the method or field with `@Valid` you can do that by calling `methodElement.annotate(Valid.class)`

**Comment by dstepanov** [comment]
I would expect to see any of the tested annotations in the code:
```
package test;
import io.micronaut.core.annotation.Introspected;
import io.micronaut.context.annotation.Executable;
import javax.inject.Singleton;
import java.util.List;
@Introspected
@Singleton
class Test {
    public List<String> list1;
    public List<String> list2;
    @Executable
    public List<String> getList1() { return null; }
    @Executable
    public List<String> getList2() { return null; }
}
```
Or I don't understand something.

**Comment by andriy-dmytruk** [comment]
I want to annotate during compile-time, not in class file.
In the test I annotate a type argument in the generated `ClassElement` which produces incorrect behavior.

I want to use this inside `Validator` to get information whether a collection needs validation (since it is not immediately verifiable whether a `Map<String, @Valid A>` needs validation).
I can however not do this and recurse through type arguments to check if any needs validation inside the `Validator`. This will mean that validation will work slightly slower for collections due to this step, but possibly this is acceptable.

**Comment by dstepanov** [comment]
Ah I see

**Comment by ojebari-ma** [comment]
Can you please confirm that this is still an issue in the latest version of Micronaut?

**Comment by dstepanov** [comment]
I believe this is fixed
